src/file_system.rs

```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2017-present, PingCAP, Inc. Licensed under Apache-2.0.
+
+use std::io::{Read, Result, Seek, Write};
+use std::path::Path;
+use std::sync::Arc;
+
+/// FileSystem
+pub trait FileSystem: Send + Sync {
+    type Handle: Send + Sync + Handle;
+    type Reader: Seek + Read + Send + ReadExt;
+    type Writer: Seek + Write + Send + WriteExt;
+
+    fn create<P: AsRef<Path>>(&self, path: P) -> Result<Self::Handle>;
+    fn open<P: AsRef<Path>>(&self, path: P) -> Result<Self::Handle>;
+    fn new_reader(&self, handle: Arc<Self::Handle>) -> Result<Self::Reader>;
+    fn new_writer(&self, handle: Arc<Self::Handle>) -> Result<Self::Writer>;
+}
+
+pub trait Handle {
+    fn truncate(&self, offset: usize) -> Result<()>;
```

Comment:
Is this method used? It's duplicated with `WriteExt`.

Reply:
`recover_queue` function in pipe_builder.rs uses it.
``` rust
  {
      warn!("The tail of raft log is corrupted but ignored: {}", e);
      f.fd.truncate(reader.valid_offset())?;
      break;
  }
```

Reply:
Okay, let's keep it for now.